pkg/identities/service.go

```diff
@@ -240,3 +249,514 @@ func NewService(kratos kClient.IdentityAPI, authz AuthorizerInterface, tracer tr
+			// needs to be validated as schemas might differ
+			Traits: traits,
+		},
+	)
+
+	// TODO @shipperizer enhance Identity resource with Permissions and Roles on the next iteration
+	// this requires calls to openfga in here unless we enhance the PrincipalContext and let that do
+	// the calls
+	if err != nil {
+		return nil, v1.NewUnknownError(err.Error())
+	}
+
+	if len(ids.Identities) != 1 {
+		return nil, v1.NewInvalidRequestError("no identity created")
+	}
+
+	return &resources.Identity{
+		Email:     identity.Email,
+		FirstName: identity.FirstName,
+		LastName:  identity.LastName,
+		Id:        &ids.Identities[0].Id,
+	}, nil
+}
+
+// GetIdentity returns a single Identity.
+func (s *V1Service) GetIdentity(ctx context.Context, identityId string) (*resources.Identity, error) {
+	ctx, span := s.core.tracer.Start(ctx, "identities.V1Service.GetIdentity")
+	defer span.End()
+
+	ids, err := s.core.GetIdentity(ctx, identityId)
+
+	if err != nil {
+		return nil, v1.NewUnknownError(err.Error())
+	}
+
+	if ids.Identities == nil || len(ids.Identities) != 1 {
+		return nil, v1.NewNotFoundError("identity not found")
+	}
+
+	id := ids.Identities[0]
+
+	traits, ok := id.Traits.(map[string]string)
+
+	if !ok {
+		traits = make(map[string]string)
+	}
+
+	// TODO @shipperizer enhance Identity resource with Permissions and Roles on the next iteration
+	// this requires calls to openfga in here unless we enhance the PrincipalContext and let that do
+	// the calls
+	i := new(resources.Identity)
+
+	i.Id = &id.Id
+
+	if email, ok := traits["email"]; ok {
+		i.Email = email
+	}
+
+	fullname, ok := traits["name"]
+	if !ok {
+		return i, nil
+	}
+
+	surnameIndex := strings.LastIndex(fullname, " ")
+
+	if surnameIndex > 0 {
+		name := strings.Trim(fullname[0:surnameIndex], " ")
+		surname := strings.Trim(fullname[surnameIndex:], " ")
+
+		i.FirstName = &name
+		i.LastName = &surname
+	}
+
+	return i, nil
+}
+
+// UpdateIdentity updates an Identity.
+func (s *V1Service) UpdateIdentity(ctx context.Context, identity *resources.Identity) (*resources.Identity, error) {
+	_, span := s.core.tracer.Start(ctx, "identities.V1Service.UpdateIdentity")
+	defer span.End()
+
+	if identity == nil {
+		return nil, v1.NewRequestBodyValidationError("bad identity payload")
+	}
+
+	traits := make(map[string]interface{})
+
+	traits["email"] = identity.Email
+	if identity.FirstName != nil && identity.LastName != nil {
+		traits["name"] = fmt.Sprintf("%s %s", *identity.FirstName, *identity.LastName)
+	}
+
+	body := kClient.NewUpdateIdentityBodyWithDefaults()
+	body.SetTraits(traits)
+
+	ids, err := s.core.UpdateIdentity(
+		ctx,
+		*identity.Id,
+		// TODO @shipperizer the code below assumes each schema has name and email
+		// needs to be validated as schemas might differ
+		body,
+	)
+
+	if err != nil {
+		return nil, v1.NewUnknownError(err.Error())
+	}
+
+	if len(ids.Identities) != 1 {
+		return nil, v1.NewInvalidRequestError("no identity created")
```

Comment:
A very trivial thing: `no identity updated`.

And is it ok/better to check `len(ids.Identities) == 0`?

Reply:
issue is that the Service v0 returns always a list, in this case the result needs to be 1 or nothing